### DISCUSSION TITLE
How do I retain all HTML, and not have it escaped?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [remarkjs/remark#922] [Q&A] [answered]
Hello! I have a simple document that I'm trying to render:

```js
import {unified} from 'unified'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import rehypeRaw from 'rehype-raw'
import rehypeSanitize from 'rehype-sanitize'
import rehypeStringify from 'rehype-stringify'

// npm init -y && npm i --save-dev node@16 && npm config set prefix=$(pwd)/node_modules/node && export PATH=$(pwd)/node_modules/node/bin:$PATH
let code = `
**bold**

<Component>
  content
  <Nested::Component @foo="foo" @bar="bar" />

  <Nested as |foo|>
    content 
  </Nested>
  <Namespaced::Component>
     <:block>
        named block/slot content
     </:block>
  </Namespaced::Component>
  <this.passedComponent as |yielded|>
     stuff {{yielded}}
  </this.passedComponent>
</Component>
`;


main()


async function main() {
  const file = await unified()
    .use(remarkParse)
    .use(remarkRehype, {sanitize: false, allowDangerousHtml: true})
    //.use(rehypeRaw)
    .use(rehypeStringify, { allowDangerousHtml: true})
    .process(code)

  console.log(String(file))
}
```
it should mostly look the same except for the bolded text at top should become HTML.
unfortunately, what happens is this: 
```
<p><strong>bold</strong></p>
<Component>
  content
  <Nested::Component @foo="foo" @bar="bar" />
<p>&#x3C;Nested as |foo|>
content
</Nested>
<a href="Namespaced::Component">Namespaced::Component</a>
&#x3C;:block>
named block/slot content
&#x3C;/:block>
&#x3C;/Namespaced::Component>
&#x3C;this.passedComponent as |yielded|>
stuff {{yielded}}
&#x3C;/this.passedComponent>
</Component></p>
```

what's goofy is that if I remove:
```
  <Nested::Component @foo="foo" @bar="bar" />
```

then this is the output:
```
<p><strong>bold</strong></p>
<Component>
  content
  <Nested as |foo|>
    content 
  </Nested>
  <Namespaced::Component>
     <:block>
        named block/slot content
     </:block>
  </Namespaced::Component>
  <this.passedComponent as |yielded|>
     stuff {{yielded}}
  </this.passedComponent>
</Component>
```

it may actually be related to the line break..?

yeah, with no line breaks in the original `code`, I get this:
```
<p><strong>bold</strong></p>
<Component>
  content
  <Nested::Component @foo="foo" @bar="bar" />
  <Nested as |foo|>
    content 
  </Nested>
  <Namespaced::Component>
     <:block>
        named block/slot content
     </:block>
  </Namespaced::Component>
  <this.passedComponent as |yielded|>
     stuff {{yielded}}
  </this.passedComponent>
</Component>
```
which is 100% correct.

So I guess... how do I keep line breaks from causing extra escaping?

**Comment by ChristianMurphy** [comment, marked as answer]
A series of things are happening here:
1. the "HTML" in your document, isn't actually HTML, it's some form of domain specific language/templating language
2. But `<Component>` is HTML like enough, that it tricks the parser into entering an HTML block
3. Inside an HTML document when a blank newline is encountered the parser switches back to markdown mode https://spec.commonmark.org/0.30/#html-blocks
4. when it gets to `<Nested as |foo|>` the `|` prevents it from being recognized as HTML, and is instead treated as a paragraph
5. when it gets to `<Namespaced::Component>` the `::` prevent that block from being recognized as HTML, and is instead treated as a paragraph
6. when it gets to `<this.passedComponent as |yielded|>` both the `.` and the `|` prevent it from being recognized as HTML, and is instead treated as a paragraph


The easy options to resolve could include:
* don't put a empty newlines, so the block is treated as raw HTML https://spec.commonmark.org/0.30/#raw-html
* use a templating language which is also valid HTML (for remark, or JSX for mdx)

Some not-as-easy options include:
* hack/customize micromark so it only has raw HTML blocks
* write a parser extension for the domain specific/templating language you are using

**Reply by NullVoxPopuli** [reply]
Thanks for the explanation on all that!

>  it's some form of domain specific language/templating language

yeah, it's a superset of HTML

> don't put a empty newlines, so the block is treated as raw HTML https://spec.commonmark.org/0.30/#raw-html

So, I could def give a go at removing newlines, since I'm already _lightly_ parsing user input.

> use a templating language which is also valid HTML (for remark, or JSX for mdx)

JSX is not an option , and also not HTML (I'm not using React, and or anything similar), _but_, my language *is* a superset of HTML.

> hack/customize micromark so it only has raw HTML blocks
> write a parser extension for the domain specific/templating language you are using

a parser extension seems intriguing. I take issue with some of the rules for escaping to/from markdown between HTML fragments (mostly with indentation), so I kinda want to pursue this anyway. How would I go about doing a parser extension or changing some of the parsing rules?

**Reply by ChristianMurphy** [reply]
> a parser extension seems intriguing. I take issue with some of the rules for escaping to/from markdown between HTML fragments (mostly with indentation), so I kinda want to pursue this anyway. How would I go about doing a parser extension or changing some of the parsing rules?

https://github.com/micromark/micromark#extending-markdown through to https://github.com/micromark/micromark#creating-a-micromark-extension describes creating a micromark extension. That creates a token stream, then an mdast-util can be created to turn that into an mdast node (or several nodes), then that new node type needs a handler in remark-rehype.
